### PR TITLE
refactor: migrate packages into additional-components

### PR DESCRIPTION
**Pull request by michealroberts**
refactor: migrate packages into additional-components

# 🚀 What's changed?

- Migrated Background package into @vue-flow/core additional components
- Migrated Controls package into @vue-flow/core additional components
- Migrated MiniMap package into @vue-flow/core additional components
- Migrated NodeResizer package into @vue-flow/core additional components
- Migrated NodeToolbar package into @vue-flow/core additional components
- Migrated PathfindingEdge package into @vue-flow/core additional components

The migration is backwards compatible, meaning imports from e.g.,:

```ts
import { Background } from '@vue-flow/background'
```

will still work. 

However, the user will see a deprecation notice.

The current vite app has had zero changes, yet still works as before:

CleanShot 2023-11-17 at 14 45 58

# Misc 🧷 

This MR includes a few smaller refactors to gather type import / exports with variable imports / exports, e.g.,:

Before:

```ts
import { zoom, zoomIdentity } from 'd3-zoom'
import type { D3ZoomEvent } from 'd3-zoom'
```

After:

```ts
import { type D3ZoomEvent, zoom, zoomIdentity } from 'd3-zoom'
```

### CODE REVIEW

**Comment by michealroberts** [comment]
@braks The motive for this PR is that we can:

- Provide Vue 2 backwards compatibility in the not to distant future
- Improve discoverability of features
- Make utilisation of xyflow/system easier in the future
- Align with xyflow/react and xyflow/svelte

**Comment by michealroberts** [comment]
> I think we can remove the pathfinding edge from the migration - it's gonna be eliminated soon anyway
> 
> It's not really used, it's not updated to support the latest Vue Flow version neither and I don't have the time to maintain it either 😄

Ok cool. I'll  leave the package as is in `@vue-flow/pathfinding-edge`.

File: packages/core/src/additional-components/Background/Background.vue [hunk 1]
```diff
@@ -1,9 +1,10 @@
 <script lang="ts" setup>
-import { useVueFlow } from '@vue-flow/core'
 import { computed, toRef } from 'vue'
-import { BackgroundVariant } from './types'
-import type { BackgroundProps } from './types'
+
 import { DefaultBgColors, DotPattern, LinePattern } from './patterns'
+import { type BackgroundProps, BackgroundVariant } from './types'
+
+import { useVueFlow } from '~/composables'
```
Comment (bcakmakoglu) [inline packages/core/src/additional-components/Background/Background.vue:7]: There's a problem with using tsconfig shortcut paths inside vue component, as they're not resolved to the actual paths when the build is triggered.

This works in `.ts` files thanks to `ts-transform-paths` but, alas, not in `.vue` files -.-

So this sadly has to be `../../../composables/useVueFlow`

Comment (bcakmakoglu) [inline packages/core/src/additional-components/Background/Background.vue:7]: There's a problem with using tsconfig shortcut paths inside vue component, as they're not resolved to the actual paths when the build is triggered.

This works in `.ts` files thanks to `ts-transform-paths` but, alas, not in `.vue` files -.-

So this sadly has to be `../../../composables/useVueFlow`

File: packages/core/src/additional-components/Controls/Controls.vue [hunk 2]
```diff
@@ -9,6 +8,10 @@ import FitView from './icons/fitview.svg'
 import Lock from './icons/lock.svg'
 import Unlock from './icons/unlock.svg'
 
+import { useVueFlow } from '~/composables'
+import { Panel } from '~/components'
+import { PanelPosition } from '~/types'
```
Comment (bcakmakoglu) [inline packages/core/src/additional-components/Controls/Controls.vue:13]: See my previous comment about shortcut paths

File: packages/core/src/additional-components/MiniMap/MiniMap.vue [hunk 3]
```diff
@@ -1,13 +1,16 @@
...
+import { pointer, select } from 'd3-selection'
+
+import { type MiniMapEmits, type MiniMapNodeFunc, type MiniMapProps, MiniMapSlots, type ShapeRendering } from './types'
 import MiniMapNode from './MiniMapNode'
-import { MiniMapSlots } from './types'
+
+import { Panel } from '~/components'
+import { useVueFlow } from '~/composables'
+import { type CoordinateExtent, type GraphNode, type PanelPosition } from '~/types'
+import { getBoundsofRects, getConnectedEdges, getRectOfNodes } from '~/utils'
```
Comment (bcakmakoglu) [inline packages/core/src/additional-components/MiniMap/MiniMap.vue:13]: Same here

File: packages/core/src/additional-components/NodeResizer/NodeResizer.vue [hunk 4]
```diff
@@ -1,11 +1,18 @@
...
+import {
+  type ControlLinePosition,
+  type ControlPosition,
+  type NodeResizerEmits,
+  type NodeResizerProps,
+  ResizeControlVariant,
+} from './types'
+import { useVueFlow } from '~/composables'
+import { NodeId as NodeIdInjection } from '~/context'
+import { type NodeDimensionChange } from '~/types'
```
Comment (bcakmakoglu) [inline packages/core/src/additional-components/NodeResizer/NodeResizer.vue:15]: And here 😄

File: packages/core/src/additional-components/NodeResizer/ResizeControl.vue [hunk 5]
```diff
@@ -1,12 +1,16 @@
...
 import { select } from 'd3-selection'
 import { drag } from 'd3-drag'
-import { ref, toRef, watchEffect } from 'vue'
+
 import type { NodeResizerEmits, ResizeControlProps, ResizeControlVariant, ResizeDragEvent } from './types'
 import { DefaultPositions, StylingProperty, getDirection } from './utils'
 
+import { useGetPointerPosition, useVueFlow } from '~/composables'
+import { type NodeChange, type NodeDimensionChange, type NodePositionChange } from '~/types'
+import { clamp } from '~/utils'
```
Comment (bcakmakoglu) [inline packages/core/src/additional-components/NodeResizer/ResizeControl.vue:12]: And also here

File: packages/core/src/additional-components/NodeToolbar/NodeToolbar.vue [hunk 6]
```diff
@@ -1,10 +1,11 @@
...
-import { NodeIdInjection, Position, getRectOfNodes, useVueFlow } from '@vue-flow/core'
+import { type CSSProperties, computed, inject, toRef } from 'vue'
+import { type Align, type NodeToolbarProps } from './types'
 
-import type { CSSProperties } from 'vue'
-import type { Align, NodeToolbarProps } from './types'
+import { useVueFlow } from '~/composables'
+import { NodeId as NodeIdInjection } from '~/context'
+import { type GraphNode, Position, type Rect, type ViewportTransform } from '~/types'
+import { getRectOfNodes } from '~/utils'
```
Comment (bcakmakoglu) [inline packages/core/src/additional-components/NodeToolbar/NodeToolbar.vue:8]: Here as well

File: packages/core/src/additional-components/index.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+export * from './Background'
+export * from './Controls'
+export * from './MiniMap'
+export * from './NodeResizer'
+export * from './NodeToolbar'
+export * from '../../../pathfinding-edge/src'
```
Comment (bcakmakoglu) [inline packages/core/src/additional-components/index.ts:6]: This line can be removed

**Review by bcakmakoglu — CHANGES_REQUESTED** [review]

File: packages/core/src/additional-components/Controls/Controls.vue (hunk 2 above)
Comment (michealroberts) [reply packages/core/src/additional-components/Controls/Controls.vue]: Amended ✅

File: packages/core/src/additional-components/MiniMap/MiniMap.vue (hunk 3 above)
Comment (michealroberts) [reply packages/core/src/additional-components/MiniMap/MiniMap.vue]: Amended ✅

File: packages/core/src/additional-components/NodeResizer/NodeResizer.vue (hunk 4 above)
Comment (michealroberts) [reply packages/core/src/additional-components/NodeResizer/NodeResizer.vue]: Amended ✅

File: packages/core/src/additional-components/NodeResizer/ResizeControl.vue (hunk 5 above)
Comment (michealroberts) [reply packages/core/src/additional-components/NodeResizer/ResizeControl.vue]: Amended ✅

File: packages/core/src/additional-components/index.ts (hunk 7 above)
Comment (michealroberts) [reply packages/core/src/additional-components/index.ts]: Done ✅

File: packages/core/src/additional-components/NodeToolbar/NodeToolbar.vue (hunk 6 above)
Comment (michealroberts) [reply packages/core/src/additional-components/NodeToolbar/NodeToolbar.vue]: Amended ✅

File: packages/core/src/additional-components/NodeResizer/ResizeControl.vue (hunk 5 above)
Comment (michealroberts) [reply packages/core/src/additional-components/NodeResizer/ResizeControl.vue]: Amended ✅

**Comment by bcakmakoglu** [comment]
Did you try to lint before comitting?

```ts
import { type PanelPosition } from '...'
```

This should give you an error by ESLint/TS as `PanelPosition` is an enum and used as an enum (hence we get an error when trying to build in this branch and the docs can't be deployed)

**Comment by michealroberts** [comment]
> Did you try to lint before comitting?
> 
> ```ts
> import { type PanelPosition } from '...'
> ```
> 
> This should give you an error by ESLint/TS as `PanelPosition` is an enum and used as an enum (hence we get an error when trying to build in this branch and the docs can't be deployed)

Yeh, 7 our of 7 successful. I think maybe it's not picking up on it because it is also deprecated.

Regardless, I have amended.

**Comment by michealroberts** [comment]
@bcakmakoglu Should I also amend docs about the usage going forward, and any deprecation notices?

**Comment by bcakmakoglu** [comment]
Mh... makes sense I'd say, otherwise we'll be stuck having to update the docs at the end and nobody will remember what has to be changed 😄

**Review by bcakmakoglu — APPROVED** [review]
Except for the docs part, the PR looks good to me :)

**Comment by michealroberts** [comment]
> Except for the docs part, the PR looks good to me :)

Nice. Will update the docs ASAP.

**Comment by bcakmakoglu** [comment]
Didn't mean to close this Pr :/
Dropped the old vue2 branch since it started to heavily conflict (there are some features that were added still 😅)

I'll take the changes from here and open a new PR to move the components into the core pkg

**Pull request closed without merging**